[test.py] Add resource consumption metrics

xtrey opened the pull request:
This PR adds the possibility to gather resource consumption metrics. The collected metrics can be used to compare performance before and after specific changes aimed at increasing performance. Currently, this functionality works only in manual mode and this is just raw data. Later on, these metrics can be used in Jupyter notebook to analyze and visualize how the resources are used and can provide the insight on how to improve it. This [PR](https://github.com/scylladb/scylladb/pull/19832) is a first insight after gathering these metrics.

Add the possibility to gather resource consumption for the test.py execution. SQLite DB will be created with different performance metrics that will allow compare the resource consumption between changes.
The DB will be in the tmp directory that by default set to testlog. Across the runs, the DB will not be deleted, so each new run will just add information to the existing DB.
Parameter --get-metrics was added to switch on or off the metrics gathering. By default, it's switched on. 
The current limitation is that the cgroup statistics is done for tests itself without clusters because even that cluster is created by the test the architecture with cluster manager leaves them in the initial cgroup.

For example, to get the top 5 tests by memory peak, we can use the next query:
```SQL
SELECT t.test_name, t.mode, t.directory, m.memory_peak
FROM tests t
JOIN test_metrics m ON t.id = m.test_id
ORDER BY m.memory_peak DESC
LIMIT 5;
```
Here is the output example:
| test\_name | mode | directory | memory\_peak |
| :--- | :--- | :--- | :--- |
| test\_read\_repair | debug | topology\_custom | 119078912 |
| test\_change\_rpc\_address | debug | topology\_custom | 81608704 |
| test\_mv\_backlog | debug | topology\_custom | 76627968 |
| test\_replace | dev | topology\_custom | 59293696 |
| test\_mv\_backlog | dev | topology\_custom | 58101760 |


Closes: https://github.com/scylladb/qa-tasks/issues/1666

Closes: https://github.com/scylladb/qa-tasks/issues/1707

mykaul commented:
Out of curiosity (as I did not follow / missed any discussion on the design of this): why not just set up a Prometheus node exporter and gather the data via it to a central Prometheus?

xtrey commented:
> Out of curiosity (as I did not follow / missed any discussion on the design of this): why not just set up a Prometheus node exporter and gather the data via it to a central Prometheus?

We don't need it for every run, it will be mostly used for local runs. So using Prometheus, I think it's overkill. Additional use case of cgroups is the possibility to control the resources and simulation of lack of CPU, memory for the Scylla process under test.

xtrey commented:
> Using Prometheus (which we have already for our on-prem hosts, btw) is an overkill, but setting a whole DB for it + collecting ourselves (in Python) the metrics in a non-standard way is better?

I'm using the SQLite db. It will be stored in testlog file as an artifact, no need to set up any new DB. Python has a built-in module to work with SQLite. AFAIK, there is not so easy to get the resources consumed in cgroup only.

xtrey commented:
@mykaul here the [doc](https://docs.google.com/document/d/1iQr4qPo8JWH43mfIoxauerWMYc0tzbUamrRFH375Vbw/edit#heading=h.prspeijbjau9) with discussion

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10550/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10550/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10550/)
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2736/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10757/)
#### Failed Tests (1/5621):
- [test_hintedhandoff_decom](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10550/testReport/junit/hintedhandoff_additional_test/TestHintedHandoff/Tests___dtest_with_topology_changes___test_hintedhandoff_decom) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_hintedhandoff_decom)
#### Build Details:
- Duration: 2 hr 31 min
- Builder: spider4.cloudius-systems.com

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10563/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10563/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10563/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2754/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10774/)
#### Failed Tests (1/5626):
- [test_ignore_dead_nodes_for_whole_dc_replace_option](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10563/testReport/junit/repair_based_node_operations_test/TestRepairBasedNodeOperations/Tests___dtest___test_ignore_dead_nodes_for_whole_dc_replace_option) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_ignore_dead_nodes_for_whole_dc_replace_option)
#### Build Details:
- Duration: 2 hr 20 min
- Builder: spider6.cloudius-systems.com

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10601/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10601/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10601/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10810/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2788/)
#### Failed Tests (2/5616):
- [test_ignore_dead_nodes_for_whole_dc_replace_option](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10601/testReport/junit/repair_based_node_operations_test/TestRepairBasedNodeOperations/Tests___dtest___test_ignore_dead_nodes_for_whole_dc_replace_option) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_ignore_dead_nodes_for_whole_dc_replace_option)
- [test_rolling_cluster_replacement_sequentially_dead_nodes_remove_and_add_multi_dc](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10601/testReport/junit/cluster_replacement_test/TestClusterReplacement/Tests___dtest___test_rolling_cluster_replacement_sequentially_dead_nodes_remove_and_add_multi_dc) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_rolling_cluster_replacement_sequentially_dead_nodes_remove_and_add_multi_dc)
#### Build Details:
- Duration: 3 hr 20 min
- Builder: spider2.cloudius-systems.com

temichus commented on test/pylib/resource_gather.py:
```diff
@@ -0,0 +1,160 @@
+        self.sqlite_writer.write_row(metrics_dict, METRICS_TABLE)
+
+
+def _parse_cpu_stat(file: TextIO, metrics: TestMetricRecord):
```
can be static method of the class

temichus commented on test.py:
```diff
@@ -1704,19 +1717,22 @@ async def main() -> int:
 
     try:
         await run_all_tests(signaled, options)
+        stop_event.set()
```
you need to wait for monitor_resources to finish as well as it has 2 seconds of sleep in the loop. also if it crashes you will be able to get stack trace only at the join

temichus commented on test/pylib/resource_gather.py:
```diff
@@ -0,0 +1,160 @@
+CGROUP_TESTS: Path = CGROUP_INITIAL.parent / 'tests'
+
+
+@define
```
can't see the advantage of using @define here. 

regular def __init__ will be more readable for me

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10634/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10634/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10634/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10837/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2816/)
#### Build Details:
- Duration: 3 hr 14 min
- Builder: i-06354bd2aee2f9bf0 (m5ad.8xlarge)

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10649/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10649/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10850/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2829/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10649/)
#### Build Details:
- Duration: 5 hr 34 min
- Builder: i-037a9a1e8658c93a7 (m5ad.8xlarge)

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10678/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10678/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10866/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10678/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2845/)
#### Build Details:
- Duration: 3 hr 14 min
- Builder: spider7.cloudius-systems.com

mlitvk commented on test/pylib/resource_gather.py:
```diff
@@ -0,0 +1,166 @@
+        os.makedirs(self.cgroup_path, exist_ok=True)
+
+    def put_process_to_cgroup(self, pid):
+        try:
+            with open(self.cgroup_path / 'cgroup.procs', "a") as cgroup:
+                cgroup.write(str(pid))
+        except ProcessLookupError:
+            logging.getLogger().log(
+                level=logging.WARNING,
+                msg=f"Process finished too quick. No metrics will be logged",
+            )
```
If the process doesn't exist, would this code trigger a `ProcessLookupError`?
I tried testing this and I got `OSError`

xtrey replied:
Double-checked and got `ProcessLookupError`. If you have `OSError`, please share the details here or in Slack. Maybe you are hitting some edge case and I need to handle it too.

avikivity commented:
> Add the possibility to gather resource consumption for the test.py execution. SQLite DB will be created with different performance metrics that will allow compare the resource consumption between changes.
> 
> Closes: [scylladb/qa-tasks#1666](https://github.com/scylladb/qa-tasks/issues/1666)
> 
> Closes: [scylladb/qa-tasks#1707](https://github.com/scylladb/qa-tasks/issues/1707)

Again, you need to explain why all of this is a good idea. Not just what we're doing.

avikivity commented on test/pylib/db/__init__.py:
Why add an empty file?

avikivity replied:
Ok. Please add license/copyright blurbs.

xtrey replied:
Added in the files where we have code. Do you want me to add it to this empty file as well?

avikivity commented on test/pylib/db/model.py:
```diff
@@ -0,0 +1,26 @@
+from attrs import define, AttrsInstance
```
Missing: copyright/license blurb.

xtrey replied:
done

avikivity commented on test/pylib/db/model.py:
```diff
@@ -0,0 +1,26 @@
+    run_id: int = None
+    directory: str = None
+    user_usec: int = None
+    system_usec: int = None
```
Suggest to use floats with units of seconds, it's more natural.

xtrey replied:
done

avikivity commented on test/pylib/db/model.py:
```diff
@@ -0,0 +1,26 @@
+    system_usec: int = None
+    usage_usec: int = None
+    memory_peak: int = None
+    time_taken: int = None
```
float

avikivity commented on test/pylib/db/writer.py:
```diff
@@ -0,0 +1,98 @@
+import sqlite3
```
copyright/license blurb.

Does this need an update to install-dependencies.sh?

xtrey replied:
There is no need to update it. The sqlite3 module is a part of the standard python library.

avikivity replied:
Nice

avikivity commented on test/pylib/resource_gather.py:
```diff
@@ -0,0 +1,166 @@
+import asyncio
```
copyright/license

xtrey replied:
done

avikivity commented on test/pylib/resource_gather.py:
```diff
@@ -0,0 +1,166 @@
+                    "sudo",
+                    "mount",
+                    "-o",
+                    "remount,rw,memory_recursiveprot,nsdelegate",
```
Please add some comments about these options.

You're affecting the callers environment. Is this okay?

xtrey replied:
Removed these options as redundant in our case. 
This intended to be used in containers. Without toolchain this part will break.

avikivity commented on tools/toolchain/dbuild:
```diff
@@ -155,7 +155,9 @@ if [ -z "$is_podman" ]; then
     docker_common_args+=(
-        --userns keep-id)
+        --userns keep-id
+        --systemd=always
```
>               • Podman sets the default stop signal to SIGRTMIN+3.

Please verify that ctrl-C still works during build and test.py execution. If not we may need to override --stop-signal.

xtrey replied:
It's not working on master branch either. Tried to add `--stop-signal=SIGTERM` and `--stop-signal=SIGKILL`, but still no luck.

avikivity replied:
It's working for me with podman, but only interactive mode (invoking dbuild without parameters).

xtrey replied:
It works when I'm already in interactive mode, when I start the build or test.py and then cancel them.

avikivity commented:
> > Add the possibility to gather resource consumption for the test.py execution. SQLite DB will be created with different performance metrics that will allow compare the resource consumption between changes.
> > Closes: [scylladb/qa-tasks#1666](https://github.com/scylladb/qa-tasks/issues/1666)
> > Closes: [scylladb/qa-tasks#1707](https://github.com/scylladb/qa-tasks/issues/1707)
> 
> Again, you need to explain why all of this is a good idea. Not just what we're doing.

Also, add an example SELECT ... LIMIT with sorting by (say) CPU consumption as a demonstration of the benefit and of how to use the new feature.

temichus reviewed (commented):
now test.py looks good. no if or for just additional method calls

ResourceGather is a real module that can be turned on/off.

spawned watcher properly finished

xtrey commented on test/pylib/resource_gather.py:
```diff
@@ -0,0 +1,166 @@
+    timeline_record = ResourceTimelineRecord()
+    while not cancel_event.is_set() and not stop_event.is_set():
+        timeline_record.cpu = psutil.cpu_percent(interval=1)
+        timeline_record.memory = psutil.virtual_memory().percent
```
We are recording the consumption of memory in the system and CPU utilization in the system. For the cgroup we are recording the peak memory consumption and CPU usage in seconds. After discussion with @kostja @temichus we decided that this will be enough, at least as an initial step. We can already gather information better than we had, however, still not as granular as can be.
I have some concerns about monitoring memory consumption in cgroup. cgroup provides a controller to get current memory consumption, that means we need to spawn monitor process equal to threads we are executing. With boost tests, it's mostly not possible. Boost tests are executed by function and many of them last less than a second, so it will be difficult to get some meaningful data. Long-lasting tests should not be affected, but still, I think it's a matter of further thorough investigation.

avikivity replied:
> We are recording the consumption of memory in the system and CPU utilization in the system. For the cgroup we are recording the peak memory consumption and CPU usage in seconds. After discussion with @kostja @temichus we decided that this will be enough, at least as an initial step. 

Maybe it's enough, but I don't understand what you're recording.

> We can already gather information better than we had, however, still not as granular as can be. I have some concerns about monitoring memory consumption in cgroup. cgroup provides a controller to get current memory consumption, that means we need to spawn monitor process equal to threads we are executing.

I'm sure you can read the cgroup report from outside the cgroup.

>  With boost tests, it's mostly not possible. Boost tests are executed by function and many of them last less than a second, so it will be difficult to get some meaningful data. Long-lasting tests should not be affected, but still, I think it's a matter of further thorough investigation.

There's getrusage().ru_maxrss. This works for a terminated process.

avikivity replied:
> > Maybe it's enough, but I don't understand what you're recording.
> 
> I'll update the docstring to the method

I mean, I understand that it's a percentage of virtual memory, just not how it's useful.

> 
> > I'm sure you can read the cgroup report from outside the cgroup.
> 
> It's not a problem to read it outside the cgroup. The issue there is that metrics exist only in real time. Once the process is done, memory will show 0 usage.


> 
> > There's getrusage().ru_maxrss. This works for a terminated process.
> 
> Please elaborate. I don't see how I can provide PID to the method to get the data and test.py is executed only once, so ru_maxrss will show maximum not only for the latest test, but for all tests that were already executed.

test.py execs ./build/mode/test/boost/whatever  and waits for it. The wait4() system call (and presumable the Python ecosystem) returns a struct rusage. The struct rusage contains ru.maxrss that says what the peak memory consumption was.

This doesn't work for forked children, but our boost tests don't fork.

It's also possible to assume some value like 50MB. If a test completes so quickly that we don't get to sample it even once, we can assume it didn't allocate a lot of memory.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10949/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10949/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10949/)
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/3073/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/11096/)
#### Failed Tests (1/31860):
- [test_group_by_with_range_name_query_paging](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10949/testReport/junit/paging_test/TestPagingData/Tests___dtest_with_topology_changes___test_group_by_with_range_name_query_paging) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_group_by_with_range_name_query_paging)
#### Build Details:
- Duration: 2 hr 14 min
- Builder: spider7.cloudius-systems.com

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10954/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10954/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/3075/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/11098/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10954/)
#### Build Details:
- Duration: 1 hr 52 min
- Builder: spider7.cloudius-systems.com

xtrey commented on test/pylib/resource_gather.py:
```diff
@@ -0,0 +1,166 @@
+    def put_process_to_cgroup(self, pid):
+        try:
+            with open(self.cgroup_path / 'cgroup.procs', "a") as cgroup:
+                cgroup.write(str(pid))
```
Rewrote to use preexec_fn

avikivity commented:
@xtrey is this ready for re-review? What are the changes?

xtrey commented:
> @xtrey is this ready for re-review? What are the changes?

Yes, it's ready.

Resolved all your comments: with int->float, usec -> sec, create pid before starting the process, reorganized the tables in DB, added monitoring for memory consumption in cgroup.

avikivity commented:
> > @xtrey is this ready for re-review? What are the changes?
> 
> Yes, it's ready.
> 
> Resolved all your comments: with int->float, usec -> sec, create pid before starting the process, reorganized the tables in DB, added monitoring for memory consumption in cgroup.

Usually, one notes the list of changes, otherwise the reviewer has to redo the review from scratch (given the scope of the changes I probably have to do that anyway).

avikivity commented on test.py:
```diff
@@ -36,6 +36,7 @@
 from test.pylib.artifact_registry import ArtifactRegistry
 from test.pylib.host_registry import HostRegistry
```
> [test.py] Add resource consumption metrics

Please enrich the patch changelog with details.

xtrey replied:
By changelog, you mean to add information to the PR description? Because I was unable to find the changelog file dedicated to test.py.

avikivity commented on test.py:
```diff
@@ -1263,7 +1280,9 @@ def report_error(error):
         except Exception as e:
-            report_error("Failed to run the test:\n{e}".format(e=e))
+            with open(f'{options.tmpdir}/stack.out', 'w+') as file:
+                traceback.print_exc(file=file)
```
I don't understand the purpose of this change.

xtrey replied:
This is leftovers after debugging. I'll revert this.

avikivity commented on test/pylib/db/model.py:
```diff
@@ -0,0 +1,42 @@
+#
+# Copyright (C) 2022-present ScyllaDB
```
It's 2024.

avikivity commented on test/pylib/db/writer.py:
```diff
@@ -0,0 +1,154 @@
+#
+# Copyright (C) 2022-present ScyllaDB
```
2024

avikivity commented on test/pylib/db/writer.py:
```diff
@@ -0,0 +1,154 @@
+        self.lock = asyncio.Lock()
+        self.conn = sqlite3.connect(database_path)
+        self.cursor = self.conn.cursor()
+        self.cursor.execute('PRAGMA foreign_keys=ON')
```
let's also issue 'PRAGMA sychronous=off' to disable fsync; fsync can stall other coroutines.

avikivity commented on test/pylib/resource_gather.py:
```diff
@@ -0,0 +1,195 @@
+
+    def get_test_metrics(self) -> Metric:
+        test_metrics: Metric = Metric(test_id=self.test_id)
+        test_metrics.time_taken = round(int(self.test.time_end - self.test.time_start))
```
Why round?

xtrey replied:
Missed to change it. It's redundant, so I'll remove it.

avikivity commented on test/pylib/resource_gather.py:
```diff
@@ -0,0 +1,195 @@
+                self.sqlite_writer.write_row(timeline_record, CGROUP_MEMORY_METRICS_TABLE)
+
+            # Control the frequency of updates, for example, every 2 seconds
+            await asyncio.sleep(1)
```
Since you have the memory.peak file, any reason to have this coroutine?

xtrey replied:
For better understanding of the memory use. For example, we have one test in a file that consumes a lot of memory, but the rest of the test do not need so much memory. With only peak memory, we will lose this information, and we will not be able to schedule the test correctly.

avikivity replied:
I don't think we'll be able to utilize this information. The execution history depends on what other tests ran in parallel. The next execution will have different timing.

xtrey replied:
Our tests executed in alphabetical order. So every run has the same order of execution. Until we start shuffling them, we can compare this data between runs.

avikivity commented on test/pylib/resource_gather.py:
```diff
@@ -0,0 +1,195 @@
+                    "/sys/fs/cgroup",
+                ],
+                check=True,
+            )
```
It's unfortunate to run so many processes for every test case, but we can tolerate it for now.

xtrey replied:
It's an architecture of the test.py. Each directory is counted as a suite, and it will have its own python process spawned. Worse situation with boost test, where each test function will have its own spawned python process. It was done so to decrease the time of the execution of the boost tests. One of the boost tests takes too much time if we execute it by file.

avikivity replied:
For boost tests, once we gather history, we can run all test cases that take a short amount of time in a single execution.

avikivity commented on test/pylib/resource_gather.py:
```diff
@@ -0,0 +1,195 @@
+        sqlite_writer.write_row(timeline_record, SYSTEM_RESOURCE_METRICS_TABLE)
+
+        # Control the frequency of updates, for example, every 2 seconds
+        await asyncio.sleep(2)
```
This is the second such coroutine, why two?

xtrey replied:
One monitoring coroutine to gather system CPU and RAM utilization, the second one is to monitor the cgroup memory usage that is spawned for each test.

avikivity replied:
I see. Do we need to keep system resource utilization in sqlite? It can't help future schedules.

xtrey replied:
It can help to analyze the whole picture of where we're still not utilizing the RAM or CPU and where we can add additional thread.

avikivity commented:
Looks reasonable. Please add an example of using sqlite3 to query the top 5 resource users and the output to the patch changelog.

xtrey commented on test.py:
```diff
@@ -1305,6 +1324,7 @@ def parse_cmd_line() -> argparse.Namespace:
         help="""Path to temporary test data and log files. The data is
         further segregated per build mode. Default: ./testlog.""",
     )
+    parser.add_argument("--gather_metrics", action=argparse.BooleanOptionalAction, default=False)
```
The main reason is to not break the flow if somebody tries to execute the test.py outside the toolchain. Additionally, I doubt that is beneficial in the current state to switch on since it just metric gathering, and they are useless if nobody analyzing them.

avikivity replied:
Why would it break the flow?

I like to keep CI using defaults, so developers aren't surprised when something works on their machine but doesn't work on CI.

We can add ~/.config/scylladb/scylladb-test-py.conf so people can adjust their configuration to taste.

xtrey replied:
The idea was to switch it on upon need. Implemented logic will not work outside the toolchain. So with default set to true and using just test.py will not work because we have no power to manipulate the cgroups in the root. Even if I rewrite the logic to start from the cgroup where the process is, there still can be an issue that this cgroup will be owned by root, and we will not be able to create another cgroup or put the PID to the desired one.

avikivity replied:
> The idea was to switch it on upon need. Implemented logic will not work outside the toolchain. 

Why not? Everything else does. Please don't write code so fragile it only works in one scenario.

> So with default set to true and using just test.py will not work because we have no power to manipulate the cgroups in the root. Even if I rewrite the logic to start from the cgroup where the process is, there still can be an issue that this cgroup will be owned by root, and we will not be able to create another cgroup or put the PID to the desired one.

I don't understand why. Podman is able to create cgroups without being root. So should we.

xtrey replied:
> Why not? Everything else does. Please don't write code so fragile it only works in one scenario.

It's not fragile. It works in the podman and in the docker. Generally speaking, in the environment where we have control. Outside the container there can be any configuration, and it will be hardly possible to cover all of them.

> I don't understand why. Podman is able to create cgroups without being root. So should we.

We can create the cgroups as well without root in the cgroups where the user is an owner. We can assume that the process where tetst.py is started located in the cgroup with correct permission, but this is not always true. In that case, we need to find the cgroup with correct permissions. After that, the flow will be actually the same. But finding the correct cgroup can be tricky a little bit.

avikivity replied:
Does this not work?

```console
$ cat /proc/$$/cgroup 
0::/user.slice/user-1000.slice/user@1000.service/app.slice/app-org.gnome.Terminal.slice/vte-spawn-aff2c052-ef9b-4023-afef-3bd36a81fa19.scope
```

xtrey replied:
It works, but for example, my foot terminal instance will be in the cgroup where the root is owner. So in most cases it will be easy to go get cgroup through the /proc/self/cgroup, but still, we cannot rely fully on it.

avikivity commented on test.py:
```diff
@@ -1776,19 +1796,24 @@ async def main() -> int:
     try:
         await run_all_tests(signaled, options)
+        stop_event.set()
+        async with asyncio.timeout(5):
```
Thanks. In Seastar we can't cancel things, so we don't have this pattern.

avikivity replied:
But, since this is just a safety mechanism, I suggest to increase the timeout even more (since we're trying to load the machine), and report when it triggered.

xtrey replied:
It's not cancelling, it's stopping of the monitoring. We are using this event as a flag that we need to stop the loop. I think this timeout is enough since it double the time of the sleep in waiting process. The only thing that I should implement is loggin an error in case of timeout happens, rather than failing the whole process. In the worst case, we will lose only the last metric with a CPU and RAM utilization, the rest should be already written.

xtrey commented on test/pylib/resource_gather.py:
```diff
@@ -0,0 +1,195 @@
+                setattr(metrics, stats[stat], float(value) / 1_000_000)
+
+
+def get_resource_gather(is_switched_on, test, tmpdir: str) -> ResourceGather:
```
I'll add an annotation to the is_switched_on, but the test will be without it. Class Test is in test.py, and it's outside python module and test.py importing the resource_gather. If I'll try to add Test in resource_gather it will lead to a recursive import, and python will raise an exception.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11115/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11115/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11115/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/164/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/3210/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/11234/)
#### Failed Tests (17/15160):
- [test_auth_v2_during_recovery.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11115/testReport/junit/%28root%29/test_auth_v2_migration/Tests___Unit_Tests___test_auth_v2_during_recovery_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_auth_v2_during_recovery.debug.1)
- [test_auth_v2_during_recovery.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11115/testReport/junit/%28root%29/test_auth_v2_migration/Tests___Unit_Tests___test_auth_v2_during_recovery_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_auth_v2_during_recovery.debug.2)
- [test_auth_v2_during_recovery.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11115/testReport/junit/%28root%29/test_auth_v2_migration/Tests___Unit_Tests___test_auth_v2_during_recovery_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_auth_v2_during_recovery.debug.3)
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11115/testReport/junit/boost.database_test.populate_from_quarantine_works.1.xunit/xml/Tests___Unit_Tests____empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11115/testReport/junit/boost.database_test.test_database_with_data_in_sstables_is_a_mutation_source_plain_fragments_monotonic_cg0.1.xunit/xml/Tests___Unit_Tests____empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11115/testReport/junit/boost.database_test.test_database_with_data_in_sstables_is_a_mutation_source_plain_fragments_monotonic_cg1.1.xunit/xml/Tests___Unit_Tests____empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11115/testReport/junit/boost.database_test.test_database_with_data_in_sstables_is_a_mutation_source_plain_read_back_cg0.1.xunit/xml/Tests___Unit_Tests____empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11115/testReport/junit/boost.database_test.test_database_with_data_in_sstables_is_a_mutation_source_plain_read_back_cg1.1.xunit/xml/Tests___Unit_Tests____empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11115/testReport/junit/boost.database_test.test_database_with_data_in_sstables_is_a_mutation_source_plain_reader_conversion_cg0.1.xunit/xml/Tests___Unit_Tests____empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11115/testReport/junit/boost.database_test.test_database_with_data_in_sstables_is_a_mutation_source_plain_reader_conversion_cg1.1.xunit/xml/Tests___Unit_Tests____empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
#### Build Details:
- Duration: 2 hr 9 min
- Builder: spider1.cloudius-systems.com

xtrey commented:
Added handling the ```asyncio.exceptions.CancelledError``` to not fail the whole job.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11142/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11142/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/11265/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/196/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/3242/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11142/)
#### Build Details:
- Duration: 6 hr 3 min
- Builder: i-03c2e03fcc7fdf4a6 (m5ad.8xlarge)

avikivity commented:
I get this on my Fedora 40 workstation:

```
$ ./test.py --mode dev
Traceback (most recent call last):
  File "/home/avi/scylla-maint/./test.py", line 2097, in workaround_python26789
    code = await main()
           ^^^^^^^^^^^^
  File "/home/avi/scylla-maint/./test.py", line 1791, in main
    setup_cgroup(options.gather_metrics)
  File "/home/avi/scylla-maint/test/pylib/resource_gather.py", line 161, in setup_cgroup
    directory.mkdir()
  File "/usr/lib64/python3.12/pathlib.py", line 1311, in mkdir
    os.mkdir(self, mode)
PermissionError: [Errno 13] Permission denied: '/sys/fs/cgroup/initial'
```

xtrey commented:
> I get this on my Fedora 40 workstation:

Added possibility to run without toolchain. However, there still be some caveats and assumptions to make it work. The terminal where we are executing, starting the process should be located in cgroup owned by the current user.

avikivity commented on test/pylib/resource_gather.py:
```diff
@@ -0,0 +1,219 @@
+    if is_required:
+        # check where the process is executed in podman or in docker
+        is_podman = os.access("/run/.containerenv", os.F_OK)
+        is_docker = os.getpid() == 1
```
I don't think this is correct. If you launch the container in interactive mode, bash will have pid 1, and test.py will have another pid. Similar if you launch it with something like ./tools/toolchain/dbuild bash -c 'ninja dev-build && ./test.py --mode dev` as I sometimes do.

xtrey replied:
It's fragile. But I found a better way, I was not aware that we switched to use by default podman instead of docker.

xtrey replied:
The behavior of cgroups is different for podman, docker and local. In docker cgroups mounted with RO permission, so we need to remount it. In podman we need to own cgroup to be able to use them how we want. Locally we cannot be user and should rely on the cgroup of the current process and hoping that user has permissions to write to it.
`container=podman` injected by systemd, so docker do not have it, and if we want to have something like it, we have to pass it manually. The best way so far that I was able to find is to check files `/run/.containerenv` for podman and `/.dockerenv`

avikivity replied:
What I meant is: check if it's mounted RO, if so remount it RW. This way, if docker/podman changes its behavior, it doesn't break dbuild.

avikivity replied:
podman/docker aren't monolithic tools. They call into runtimes, which can be different based on the OS, and the runtimes change independently of podman/docker (crun/runc/whatever).

avikivity commented on test/pylib/resource_gather.py:
```diff
@@ -0,0 +1,219 @@
+                    "remount,rw",
+                    "/sys/fs/cgroup",
+                ],
+                check=True,
```
Do the two container environments really need such different methods to give the current user permissions?

Perhaps we should revert 34e3302c017? Though the equivalent was necessary, at least with older docker.

Perhaps recent docker can remap files written as root to the original user.

xtrey replied:
Unfortunately, yes. It is still necessary to remount the cgroup for Docker. Change that you mentioned affects podman, not docker.

avikivity replied:
So here, look at /proc/mounts, check if the /sys/fs/cgroup entry is `ro`, and change it to rw if it is. This way, changes in the docker/podman runtime don't affect us.

xtrey replied:
I've modified the code to check that we in a docker container and remount only for docker. So current code does not affect local run or execution in podman.

avikivity replied:
Ok. It would be better to only check for is_cgroup_ro, not is_docker. The idea is that is docker or podman change their behavior in the future, and we build an old release, we don't have to fight with this script.

xtrey replied:
podman not allow remounting. If podman changes something here, we will need to modify our approach. This was my main point, we cannot predict changes in docker/podman so we cannot write a code before these changes will arrive. For docker remounting seems redundant, with privileged mode it will mount RW, without it will be RO, but there will be no way to remount it with sudo.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11282/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11282/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11282/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/317/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/3364/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/11388/)
#### Failed Tests (2/15326):
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11282/testReport/junit/boost.loading_cache_test.test_loading_cache_reload_during_eviction.3.xunit/xml/Tests___Unit_Tests____empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11282/testReport/junit/boost.loading_cache_test.test_prepared_statement_small_cache.3.xunit/xml/Tests___Unit_Tests____empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
#### Build Details:
- Duration: 2 hr 55 min
- Builder: i-0ff42e919984bf83f (m5ad.8xlarge)

xtrey commented:
Removed `--systemd=always` for podmand since we migrated to a newer version of Fedora and podman, so no need for this workaround for the bug https://github.com/containers/podman/issues/20073
Changed the way of distinguishing that the process in the docker container

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11391/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11391/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/392/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/3443/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/11466/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11391/)
#### Failed Tests (3/27390):
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11391/testReport/junit/raft.many_test.test_many_1000.1.xunit/xml/Tests___Unit_Tests____empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11391/testReport/junit/raft.many_test.test_many_1000.2.xunit/xml/Tests___Unit_Tests____empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11391/testReport/junit/raft.many_test.test_many_1000.3.xunit/xml/Tests___Unit_Tests____empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
#### Build Details:
- Duration: 3 hr 57 min
- Builder: i-0fa59feaebe1d1f07 (m5ad.8xlarge)

xtrey commented:
Added try-catch when stopping the cgroup monitoring.
This is flaky, and I can't reproduce it locally to check that the fix is working. Even with repeat 100 there were no log messages regarding the `CancelledError` or `TimeoutError`
![image](https://github.com/user-attachments/assets/e6aaa2b8-fc9c-4d98-ac34-1c6914e03a66)

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11413/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11413/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/410/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/3460/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/11483/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11413/)
#### Build Details:
- Duration: 3 hr 38 min
- Builder: spider1.cloudius-systems.com

xtrey commented:
Added cleaning the cgroup after the test

xtrey commented:
Remount cgroup2 will be done for docker only and when it only mounted in RO mode.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11643/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11643/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/583/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/11653/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/3638/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11643/)
#### Build Details:
- Duration: 6 hr 36 min
- Builder: i-00b8e1a2badd4478a (m5ad.8xlarge)
